{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d8589ed4-a4a7-4bb7-9d2f-410d465ca7ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import h5py\n",
    "\n",
    "class VariationDataset:\n",
    "    \"\"\"\n",
    "    Class to hold and process sequence variation and phenotypic data. For each gene, the\n",
    "    following data should be provided (via input `seq_data_paths` in at instantiation):\n",
    "        * A Pandas DataFrame of variants with columns\n",
    "            - `var_id` (as index): generally of the form chr:pos:ref:alt, but can be something else\n",
    "            - `AA_POS`, `AA_REF`, `AA_ALT`, `AA_STRAND`: the position and strand of the\n",
    "              variant, and the reference and alternate alleles\n",
    "            - Other optional information (e.g. allele counts/frequencies from VCF file)\n",
    "        * A Pandas DataFrame of patients and their sequences with columns:\n",
    "            - `patient_id` (as index): the ID of each patient\n",
    "            - `seq_id`: the ID of a patient's sequence\n",
    "            - Other optional information (e.g. the number of variants per sequence, the \n",
    "              IDs of each variant per sequence, sequence counts)\n",
    "        * A Pandas DataFrame of unique sequences with columns:\n",
    "            - `seq_id` (as index): the ID of a unique sequence\n",
    "            - `vars_list`: the variant IDs in the sequence, joined with '_', like an \n",
    "              adjacency list\n",
    "            - `n_vars`: the number of variants in the sequence\n",
    "            - `seq_count`: the count of  a unique sequence in the data \n",
    "        * The sequence data (either one-hot encoding or some embedding), saved as an\n",
    "          binary `.npy` array\n",
    "    \n",
    "    NOTE: the Pandas DataFrames are expected to be stored as `.parquet` files.\n",
    "          \n",
    "    The phenotypic data should be a Pandas DataFrame stored as a `.parquet` file, and it\n",
    "    must contain a `patient_id` column that matches the identically-named column in each\n",
    "    sequence table\n",
    "    \n",
    "    NOTE: The class can be given the variation data for one or more genes. Gene-specific  \n",
    "    suffixes will be added to the appropriate column names to reduce confusion.\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, seq_data_paths, phen_data_path):\n",
    "        \"\"\"        \n",
    "        -----------------------------------------------------------------------------------\n",
    "        Inputs          | Description\n",
    "        ----------------+------------------------------------------------------------------\n",
    "        seq_data_paths  | Two-level dictionary. First level of keys should be the sequence \n",
    "                        | (gene) names, and the second level dictionaries should be like:\n",
    "                        | {'variants': VARIANTS_TABLE_PATH, \n",
    "                        |  'sequences': SEQUENCE_TABLE_PATH, \n",
    "                        |  'patients': PATIENT_DATA_PATH, \n",
    "                        |  'seqdata': SEQUENCE_DATA_PATH}\n",
    "        ----------------+------------------------------------------------------------------\n",
    "        phen_data_path  | The path to the `.parquet` dataframe of patient phenotypes\n",
    "        -----------------------------------------------------------------------------------\n",
    "        \n",
    "        \"\"\"\n",
    "        self.seq_data_paths = seq_data_paths\n",
    "        self.phen_data_path = phen_data_path\n",
    "        self.gene_names = list(seq_data_paths.keys())\n",
    "        self.n_genes = len(self.gene_names)\n",
    "        self.many_genes = (self.n_genes > 1)\n",
    "        \n",
    "        self.variants = {s:None for s in self.gene_names}\n",
    "        self.sequences = {s:None for s in self.gene_names}\n",
    "        self.patients = {s:None for s in self.gene_names}\n",
    "        self.seqdata = {s:None for s in self.gene_names}\n",
    "        \n",
    "        # read tables and data\n",
    "        self.summary = {}\n",
    "        total_seq_length, total_n_vars = 0, 0\n",
    "        \n",
    "        for gene in self.gene_names:\n",
    "            self.variants[gene] = pd.read_parquet(seq_data_paths[gene]['variants'])\n",
    "            self.sequences[gene] = pd.read_parquet(seq_data_paths[gene]['sequences'])\n",
    "            self.patients[gene] = pd.read_parquet(seq_data_paths[gene]['patients'], \n",
    "                                                 columns = ['n_vars', 'seq_id'])\\\n",
    "                                    .add_prefix(f\"{gene}:\" if self.many_genes else \"\")\n",
    "            self.seqdata[gene] = np.load(seq_data_paths[gene]['data'])\n",
    "            \n",
    "            total_length += self.seqdata[gene].shape[1]\n",
    "            total_n_vars += len(self.variants[gene])\n",
    "            self.summary[gene] = {\"length\": self.data[gene].shape[1], \n",
    "                                  \"n_vars\": len(self.variants[gene]), \n",
    "                                  \"n_seqs\": len(self.sequences[gene])}\n",
    "        \n",
    "        if not self.many_genes:\n",
    "            self.variants = self.variants[gene]\n",
    "            self.sequences = self.sequences[gene]\n",
    "            self.patients = self.patients[gene]\n",
    "            self.data = self.data[gene]\n",
    "        else:\n",
    "            # combine patient sequences\n",
    "            self.patients = pd.concat(self.patients.values(), axis = 1)\n",
    "            # sum the number of variants per patient\n",
    "            self.patients['n_vars'] = self.patients.loc[:,self.patients.columns.str.contains('n_vars')].sum(1)\n",
    "            # combine the sequence IDs across all genes to generate new network-level sequence IDs\n",
    "            self.patients['jointed_seq_ids'] = self.patients\\\n",
    "                                                .loc[:, self.patients.columns.str.contains('seq_id')]\\\n",
    "                                                .agg('_'.join, axis = 1)\n",
    "            # create new sequences table\n",
    "            sequences = self.patients['jointed_seq_ids'].copy().drop_duplicates()\n",
    "            sequences.index.name = 'new_seq_id'\n",
    "            \n",
    "            # assign combined sequence IDs to patients\n",
    "            self.patients = pd.merge(left = self.patients.reset_index(), \n",
    "                                     right = sequences.reset_index(), \n",
    "                                     left_on = 'jointed_seq_ids', \n",
    "                                     right_on = 'new_seq_id', \n",
    "                                     how = 'left')\n",
    "            self.patients = self.patients[[\"patient_id\", \"n_vars\", \"new_seq_id\"]]\\\n",
    "                                                    .astype({'new_seq_id': 'category'})\\\n",
    "                                                    .set_index(\"patient_id\")\n",
    "            # count combined sequences\n",
    "            seq_counts = self.patients['new_seq_id'].value_counts().rename('seq_count')\n",
    "            seq_counts = seq_counts.astype('int32')\n",
    "            sequences = pd.merge(left = sequences, \n",
    "                                 right = seq_counts,\n",
    "                                 left_on = \"new_seq_id\", \n",
    "                                 right_index = True,\n",
    "                                 how = \"left\")\n",
    "            \n",
    "            # resplit the combined sequence IDs in the sequences table\n",
    "            sequences[self.gene_names] = sequences['jointed_seq_ids'].str.split('.', expand = True)\n",
    "            sequences = sequences.drop(columns = ['jointed_seq_ids'])\n",
    "            \n",
    "            self.sequences['all'] = sequences\n",
    "            \n",
    "            # add to summary\n",
    "            self.summary['all_genes'] = {'length' : total_length, \n",
    "                                         'n_vars': total_n_vars, \n",
    "                                         'n_seqs': len(sequences)}\n",
    "            \n",
    "        self.summary = pd.DataFrame(self.summary).T\n",
    "            \n",
    "        self.seq_lengths = self.summary['length'].tolist() \n",
    "        self.n_sequences = self.summary['n_seqs'].tolist() \n",
    "        if not self.many_genes: \n",
    "            self.seq_lengths, self.n_sequences = self.seq_lengths[0], self.n_sequences[-1]\n",
    "        self.seq_idxs = np.arange(0, self.n_sequences)\n",
    "                \n",
    "    def weights(self, t):\n",
    "        \"\"\"\n",
    "        Calculate the training weights of the sequences using their counts in the population. \n",
    "        Parameter t <= 1 is to flatten the distribution (smaller t leads to flatter distributions):\n",
    "        \n",
    "        w'_i = c_i / (sum_j c_j)\n",
    "        w_i = w'_i^t / (sum_j w'_j^t)\n",
    "        \n",
    "        \"\"\"\n",
    "        counts = self.sequences['seq_count'].to_numpy()\n",
    "        weights = counts/counts.sum()\n",
    "        weights = weights ** t\n",
    "        return weights/weights.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e1f9094-a81e-4303-9d1c-be1b3e36bb79",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
